zookeeper-server/src/main/java/org/apache/zookeeper/server/util/QuotaMetricsUtils.java

```diff
@@ -0,0 +1,158 @@
+    private static Map<String, Number> getQuotaLimitOrUsage(final DataTree dataTree,
+                                                            final QUOTA_LIMIT_USAGE_METRIC_TYPE type) {
+        final Map<String, Number> metricsMap = new ConcurrentHashMap<>();
+        if (dataTree != null) {
+            getQuotaLimitOrUsage(Quotas.quotaZookeeper, metricsMap, type, dataTree);
+        }
+        return metricsMap;
+    }
+
+    static void getQuotaLimitOrUsage(final String path,
+                                     final Map<String, Number> metricsMap,
+                                     final QUOTA_LIMIT_USAGE_METRIC_TYPE type,
+                                     final DataTree dataTree) {
+        final DataNode node = dataTree.getNode(path);
+        if (node == null) {
+            return;
+        }
+        final Set<String> children = node.getChildren();
+        if (children.isEmpty()) {
+            if (shouldCollect(path, type)) {
+                collectQuotaLimitOrUsage(path, node, metricsMap, type);
+            }
+            return;
+        }
+        for (final String child : children) {
+            getQuotaLimitOrUsage(path + "/" + child, metricsMap, type, dataTree);
+        }
+    }
+
+    static boolean shouldCollect(final String path, final QUOTA_LIMIT_USAGE_METRIC_TYPE type) {
+        return path.endsWith(LIMIT_END_STRING)
+                && (QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_COUNT_LIMIT == type || QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_BYTES_LIMIT == type)
+                || path.endsWith(STATS_END_STRING)
+                && (QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_COUNT_USAGE == type || QUOTA_LIMIT_USAGE_METRIC_TYPE.QUOTA_BYTES_USAGE == type);
+    }
+
+    static void collectQuotaLimitOrUsage(final String path,
+                                         final DataNode node,
+                                         final Map<String, Number> metricsMap,
+                                         final QUOTA_LIMIT_USAGE_METRIC_TYPE type) {
+        final String namespace = PathUtils.getTopNamespace(Quotas.trimQuotaPath(path));
+        if (namespace == null) {
+            return;
+        }
+        if (node.getData() == null) {
+            return;
+        }
+        final StatsTrack statsTrack = new StatsTrack(node.getData());
+        switch (type) {
+            case QUOTA_COUNT_LIMIT:
+                metricsMap.put(namespace, getQuotaLimit(statsTrack.getCountHardLimit(), statsTrack.getCount()));
+                break;
+            case QUOTA_BYTES_LIMIT:
+                metricsMap.put(namespace, getQuotaLimit(statsTrack.getByteHardLimit(), statsTrack.getBytes()));
+                break;
+            case QUOTA_COUNT_USAGE:
+                metricsMap.put(namespace, statsTrack.getCount());
+                break;
+            case QUOTA_BYTES_USAGE:
+                metricsMap.put(namespace, statsTrack.getBytes());
+                break;
```

Comment:
There is a mismatch, in general, between the notions of quota path (which cover subtrees, potentially deep in the hierarchy) and that of "top namespace."  I know you did not introduce the latter, and I understand that splitting just below the root is convenient in most deployments—but that mismatch is bound to rear its ugly head at the most inopportune moments :)

The above suffers from it, as:

```
create /ZOOKEEPER-4211
create /ZOOKEEPER-4211/a
create /ZOOKEEPER-4211/a/b
setquota -B 32 /ZOOKEEPER-4211/a/b

create /ZOOKEEPER-4211/a/z
setquota -B 128 /ZOOKEEPER-4211/a/z

set /ZOOKEEPER-4211/a/b 0123456789012345678901234567890123456789
```

causes the `/a/z` subtree to "cover" `/a/b` and report the wrong results:

```
quota_bytes_limit_per_namespace{key="ZOOKEEPER-4211",} 128.0
quota_bytes_usage_per_namespace{key="ZOOKEEPER-4211",} 0.0
```

I would suggest *summing* the values found in multiple subtrees of a given "namespace."

Reply:
Yes, good suggestion. In our case, we only set the quota on the top namespace, so no need to sum. But for general use case, we want to do sum.

Reply:
changed

Reply:
Okay!